[프레디]미션 5: 보드에 위치 부여 및 점수 계산

Dae-Hwa opened the pull request:
안녕하세요! 프레디 입니다.
미션 5 완료했습니다.

**Board**

- `initialize`시 비어있는 기물을 추가해줬습니다.
- `initializeWithSort`를 이용하여 기물들을 정렬하여 저장할 수 있도록 했습니다.
- `findPiece`와 `set`을 요구사항에 맞게 `String`으로 작동하도록 수정했습니다. 이 과정에서 `PositionConverter`를 이용하여 중복을 제거했습니다.
- 색깔과 종류 별로 몇 개가 있는지 알 수 있는 `getNumberOf`를 추가했습니다
- `ChessCalculator`에서 사용할 `groupingByCalculablePiece`를 구현했습니다.

**File**

- 이전 리뷰에서 `set`과 `get`에서 `char`인덱스를 변환하는 과정이 이해하기 어렵다는 피드백이 있어서, `PositionConverter`는 외부(`Board`)에서만 사용하도록 하고 내부의 구현은 자연스러운 인덱스를 사용하도록 수정했습니다.

- 파일에 들어있는 색상 별 기물의 수량을 알 수 있는 `getNumberOf`를 추가했습니다.

- 빈 칸으로 채워주는 `fillWithBlank`를 추가했습니다.

- 기물들을 스트림으로 변환해주는 `getPieces`와 `getCalculablePieces`를 추가했습니다. `Board`에서 사용할 때 여러 개의 스트림을 반복적으로 호출하는 구조로 되어 있어 `flatmap`을 이용하여 합쳐주기 위해 `Stream`을 리턴하도록 했습니다.

- `getCalculablePieces`는 점수 계산이 가능하도록 `Piece`를 `CalculablePiece`로 변환 해주는 작업입니다.

  > 관련 된 설명 후술

**CalculablePiece**

계산을 어떤 식으로 할지 고민을 많이 했습니다. 크게 두 가지 방법을 생각하고 했는데, `Piece`에 위치 정보를 추가하는 방법과 `Board`를 계산할 때 90도 돌려서 생각하는 방법을 고려했습니다.

일단 기본적으로 점수 계산은 게임 진행에 필수적이라기보다는 부수적인 역할로 느껴져서 하나의 기능 때문에 `Piece`에 위치 정보를 추가하는 게 적절한지에 대한 의문이 많이 들었습니다. 따라서 90도로 돌리는 방법으로 먼저 구현했는데, 생각보다 복잡도가 높았고 각 객체의 역할도 너무 중구난방이 되는 것 같았습니다.

리팩토링을 진행하며 자연스럽게 `Piece`안에 위치 정보를 추가하는 형식으로 구현이 되었는데, 하지만 앞서 말씀드렸듯이 `Piece`에 위치 정보를 추가하는 데는 회의적이어서 계산에만 사용할 수 있는 VO로 `CalculablePiece`를 새로 만들어 추가했습니다.

**Color**

- 요구 사항에 맞게 `NOCOLOR`를 추가했습니다.

**Kind**

- 요구사항에 맞게 `EMPTY`와 `point`를 추가했습니다.

**Piece**

- 빈 기물을 만드는 `createBlank`를 추가했습니다.
- 정렬을 위해 `Comparable`을 구현했습니다. 내림차순으로 정렬하기 위해 -1을 곱하는 식으로 구현했는데, 위치를 바꿔서 계산하는 것 보다 이 편이 가독성이 좋다고 판단했습니다.

**PieceFactory**

- `PieceFactory`에도 `creatBlank`를 추가해줬습니다. 상태가 필요 없기 때문에 `static`으로 구현했는데, 다른 메소드들과 일관성이 없는 것이 아닌가 하는 생각도 했는데, 지금 상황에서는 이게 맞는 것 같다고 결론을 내렸습니다.

**Calculable**

- `ChessCalculator`에서 사용할 수 있는 인터페이스입니다. `groupingByCalculablePiece`를 구현하도록 되어있습니다.

**ChessCalculator**

- `Calculable` 의 `groupingByCalculablePiece`를 이용하여 점수를 계산하도록 했습니다. 집계를 하며 `Pawn`에 대한 제약사항을 처리해줬습니다.

**PositionConverter**, **FileIndexConverter**, **RankIndexConverter**

- `fileIndex`(a\~h)와 `rankIndex`(1\~8)이 배열의 인덱스(0\~7)와 일치하지 않는 것을 해결해주기 위한 변환 클래스입니다.

감사합니다!

Dae-Hwa committed:
Add createBlank in Piece with update Color and kind

빈 칸을 표현할 수 있는 팩토리 메서드 추가. 표현을 위해 Color.NOCOLOR와 Kind.EMPTY 추가

Dae-Hwa committed:
Add createBlank in PieceFactory

생성할 때 상태가 필요없기 때문에 static으로 생성. 일관성을 헤치는 것인지 고려해봐야함.

Dae-Hwa committed:
Add fillWithBlank in File

Dae-Hwa committed:
Refactor fillWithPawn and fillWithBlank in File

fillWith(Piece) 추가하여 중복 제거

Dae-Hwa committed:
Rename initializeBy in Board to initializedPieceBy

Dae-Hwa committed:
Add getPieceNumberBy in File

Dae-Hwa committed:
Add set in File

Dae-Hwa committed:
Fix the typo of comment in File

Dae-Hwa committed:
Fix getPieceNumberBy in File and rename to getNumberOf

Kind에 따라 개수 반환하도록 수정. 기능에 맞게 메소드명 변경

Dae-Hwa committed:
Add set and getNumberOf in Board

Dae-Hwa committed:
Add toMap in File

Dae-Hwa committed:
Add initializeEmptyPiece in Board

Dae-Hwa committed:
Add point in Kind

Dae-Hwa committed:
Add getRotatedFilesBy and calculate whice need refactoring

Dae-Hwa committed:
Move calculate and getRotatedFilesBy in Board to ChessCalculator

Board에 있는 calculate와 getRotatedFilesBy를 ChessCalculator로 이동
ChessCalculator의 getRotatedFilesBy를 숨기기 위해 인터페이스 사용

Dae-Hwa committed:
Remove unnecessary imports

Dae-Hwa committed:
Refactor calculate in ChessCalculatorImpl with adding groupingByCalculablePiece

기존의 90도 돌리는 방식에서, CalculablePiece에 담아서 모은 뒤, CalculablePiece를 키로 하고, point의 합을 value로 하는 Map을 만든다.
계산하면서 pawn의 제약사항에 따라 /2를 해주도록 구현됐다.
리팩토링 결과로 책임이 적절하지 않다고 생각했던 File의 calculate가 삭제되었다.

Dae-Hwa committed:
Remove unnecessary imports

Dae-Hwa committed:
Add getPoint in Piece

Dae-Hwa committed:
Refactor ChessCalculatorImpl to use method reference

Dae-Hwa committed:
Add initializeWithSort with blackAndWihtePieces in Board

Map<Color, List<Piece>> 의 구조로 색상 별 기물을 저장할 수 있도록 추가. 기물을 가져오며 정렬해준다. 정렬은 점수가 높은 순으로 되어있다.

Dae-Hwa committed:
Update toString of Board

Dae-Hwa committed:
Sort methods in Board and BoardTest

Dae-Hwa committed:
Refactor BoardTest and FileTest with adding TestBaseOfBoardPackage

Dae-Hwa committed:
Update findPiece and set in Board to fulfill requirement

Dae-Hwa committed:
Fix comments in File that not updated with commit before

Dae-Hwa committed:
Replace set and get in File for using array index instead char index

Dae-Hwa committed:
Refactor 'findPiece' and 'set' in 'Board' to using PositionConverter that added with

PositionConverter 는 String으로 위치를 받아 리스트에 사용할 수 있도록 인덱스를 변환시켜준다.

Dae-Hwa committed:
Refactor PositionConverter with adding FileIndexConverter and RankIndexConverter

Dae-Hwa committed:
Refactor Board to use RankIndexConverter

Dae-Hwa committed:
Refactor ChessCalculatorImpl to use stream instead for loop

Dae-Hwa committed:
Change responsibility of Board and ChessCalculator

이전에 ChessCalculator를 Board 안에서 사용했는데, 이제 굳이 그럴 필요가 없으므로 밖에서 별개로 조합하여 사용할 수 있도록 수정. 수정을 위해 groupingByCalculablePiece를 Board로 옮김. 수정 후 getScoreOf 삭제

Dae-Hwa committed:
Remove magic numbers

Dae-Hwa committed:
Create Calculable to use at ChessCalculator

Dae-Hwa committed:
Remove unnecessary import

Dae-Hwa commented:
이번 미션에서는 소스 트리를 활용해봤는데, 메세지 길이가 생각과 다르게 된 부분이 생겼습니다. 관련 설정을 해줬고, 앞으로는 깔끔하게 작성하도록 하겠습니다.

Dae-Hwa committed:
Change the place of setBoard in BoardTest

LarryJung commented on src/main/java/kr/codesquad/freddie/chess/board/Board.java:
```diff
@@ -1,23 +1,66 @@
     // 체스에서 row를 rank라고 한다.
     public static final int RANK_SIZE = 8;
     private List<File> files = new ArrayList<>();
+    private Map<Color, List<Piece>> blackAndWhitePieces = new HashMap<>();
```
작은 의견인데요ㅎㅎ 이제 Color에 NOCOLOR도 있어서 blackAndWhitePieces 는 조금 안맞을...수도? 있지 않을까 해서
colorPieceMap 같은 변수명도 괜찮아 보입니다. ㅎㅎ 필수는 아니니 그렇다고만 생각해주세요~

LarryJung replied:
아 그리고, 요 blackAndWhitePieces는 꼭 필요한 멤버변수일까요? files 에 있는 데이터만으로 충분해 보이는데요~

Dae-Hwa replied:
NOCOLOR를 생각 못하고 있었네요 😓
blackAndWhitePieces는 요구사항에 각 색에 해당하는 기물을 collection에 저장하여 정렬하라고 나와있어 추가했던 것인데,  사실 저도 꼭 필요하다고 생각되지는 않았습니다. 혹시 제가 요구사항을 잘못 이해한 것일까요?

LarryJung replied:
멤버 변수가 필요 없어도 구현이 가능할 땐 왠만하면 제거하시는게 좋습니다. 데이터가 이중으로 관리되고 있다는 뜻일수도 있어서요~

Dae-Hwa replied:
감사합니다! 애매했던 부분인데 말씀해주신 대로 반영해서 수정했습니다!
9a7efdb

LarryJung commented on src/main/java/kr/codesquad/freddie/chess/board/Board.java:
```diff
@@ -1,23 +1,66 @@
+
+    private List<Piece> getPiecesBy(Color color) {
+        return files.stream()
+                .flatMap(File::getPieces)
```
flapMap 을 사용하셨네요!

LarryJung commented on src/main/java/kr/codesquad/freddie/chess/board/Board.java:
```diff
@@ -52,31 +89,31 @@ public int pieceCount() {
-    private int convertRankIndexToListIndex(int rankIndex) {
-        return RANK_SIZE - rankIndex;
+    public Map<Color, List<Piece>> getBlackAndWhitePieces() {
+        return Collections.unmodifiableMap(blackAndWhitePieces);
```
이런 불변 객체를 사용해주시는 습관은 좋습니다~

LarryJung commented on src/main/java/kr/codesquad/freddie/chess/board/File.java:
```diff
@@ -23,34 +23,39 @@ public File add(Piece piece) {
     }
 
+    public int getNumberOf(Color color, Kind kind) {
+        return (int) pieces.stream()
```
형변환 보다는 Integet.valueOf 같은 것을 사용해보세요~

Dae-Hwa replied:
Long의 intValue() 를 이용하여 수정했습니다.
8c63643

LarryJung commented on src/main/java/kr/codesquad/freddie/chess/board/File.java:
```diff
@@ -23,34 +23,39 @@ public File add(Piece piece) {
+
+    private void fillWith(Piece piece) {
         while (isAddable()) {
-            pieces.add(new Piece(color, Kind.PAWN));
+            add(piece);
```
while 보다는 if가 좀더 적절하지 않을까요?

Dae-Hwa replied:
죄송합니다. 잘 이해가 되지 않아 질문드리려 합니다.
제시해주신 방법을 아래와 같은 방식으로 이해했는데 제가 이해한 게 맞을까요?
```java
while (true) {
    if (!isAddable()) {
        throw new IllegalArgumentException();
    }
}
```

LarryJung replied:
아 저는
```java
if (isAddable) { ... }
```

이렇게 제안을 드렸습니다. ㅎㅎ

Dae-Hwa replied:
while 대신 if 를 사용하라는 말씀이신가요??

LarryJung replied:
아~ 네 그런데요, 코드를 보니 while이 의도된 것이었군요? 그렇다면 없는 코멘트로 치시죠....😅

Dae-Hwa replied:
확인 해주셔서 감사합니다!

LarryJung commented on src/main/java/kr/codesquad/freddie/chess/board/File.java:
```diff
@@ -68,6 +73,15 @@ public void fillWithRoyal(Color color) {
+        return IntStream.range(0, SIZE)
+                .mapToObj(i -> CalculablePiece.create(this.pieces.get(i), (char) (i + 'a')));
+    }
+
```
짜다보면 Stream을 반환하고자 싶은 욕구가 있지만.. Stream형태로 리턴하는 것은 안티패턴입니다.
스트림은 반환된 값을 두번이상 사용하지 못합니다. 이름도 그래서 스트림인데요~ 객체 생성비용이 조금 든다 해도 List로 반환하고 필요할 때 변환해서 사용하는 것을 권장합니다.

Dae-Hwa replied:
List를 리턴하도록 수정했습니다. 
adf665c
안티패턴이라 하셔서 수정하긴 했는데, 호출한 곳에서 collect하여 사용하는 것도 안 좋은 것인가요? 좀 더 유연해질거라 생각했는데, 제가 생각했던 거랑 실제로 사용할 때가 다른가보네요 😞

LarryJung replied:
리턴값은 Stream이 아닌 List, Map 등과 같은 일반 Collection 타입으로 지켜주시는게 좋습니다. 반환하기 위해서 Collector 를 사용하는 건 크게 문제가 되지 않습니다! Stream은 특정하게 호출한 쪽에서 요구되는 타입인 것이고 stream을 반환하는 객체 입장에서는 Stream을 반환할 이유가 별로 크지 않져..

Dae-Hwa replied:
이유가 궁금했는데 답변해주셔서 감사합니다!

LarryJung commented on src/main/java/kr/codesquad/freddie/chess/piece/CalculablePiece.java:
```diff
@@ -0,0 +1,30 @@
+
+import java.util.Objects;
+
+public class CalculablePiece extends Piece {
+    private char fileIndex;
+
+    private CalculablePiece(Color color, Kind kind, char fileIndex) {
+        super(color, kind);
+        this.fileIndex = fileIndex;
+    }
+
+    public static CalculablePiece create(Piece piece, char fileIndex) {
+        return new CalculablePiece(piece.getColor(), piece.getKind(), fileIndex);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        if (!super.equals(o)) return false;
+        CalculablePiece that = (CalculablePiece) o;
+        return fileIndex == that.fileIndex;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(super.hashCode(), fileIndex);
+    }
+}
```
이 클래스를 도출해내는 과정에서 많은 성장이 있었을 거라 생각이 드네요 ㅎㅎ
현재는 position이 점수를 계산하는 용도라고 생각이 들 수 있지만 position 정보는 체스에 꼭 필요한 요소이니 이 클래스가 좀 더 중요해질 수도 있을 것 같습니다~

LarryJung commented on src/main/java/kr/codesquad/freddie/chess/utils/Calculable.java:
```diff
@@ -0,0 +1,10 @@
+
+public interface Calculable {
+    Map<CalculablePiece, Double> groupingByCalculablePiece(Color color);
+}
```
요 인터페이스는 꼭 필요한 것일까요? calculateScore의 인자가 Board가 아닌 Calculable 인터페이스로 선언했을때 장점이 있었나요? 
ChessCalculator가 Board말고 같은 인터페이스로 선언된 다른 어떤 클래스로 수용할 수 있는 다형적인 기능을 제공한다면 이점이 있겠지만, 현재는 불필요해보입니다~

Dae-Hwa replied:
그룹으로 묶는게 Board의 역할이 아닌 것 같아 여러 방법을 생각해보다가 인터페이스로 추가적인 역할을 명시해주는 장점이 있을 것이라 생각했습니다.
말씀해주신 대로 현재는 불필요한 복잡성이 추가되는 것 같아 제거했습니다. 다음에 기억했다가 말씀해주신 상황이 되면 다시 활용해봐야겠습니다! 감사합니다
fa7d550

LarryJung commented on src/test/java/kr/codesquad/freddie/chess/board/TestBaseOfBoardPackage.java:
```diff
@@ -0,0 +1,9 @@
+import kr.codesquad.freddie.chess.piece.Color;
+import kr.codesquad.freddie.chess.piece.PieceFactory;
+
+public class TestBaseOfBoardPackage {
```
클래스 이름에 package는 조금 어색한것 같습니다. 
패키지 이름이 바뀌면 요 클래스 이름도 바뀔까요?

Dae-Hwa replied:
BoardTestBase로 이름을 변경했습니다.
de3b31c

LarryJung commented on src/main/java/kr/codesquad/freddie/chess/utils/PositionConverter.java:
```diff
@@ -0,0 +1,33 @@
+        this.rankIndexConverter = rankIndexConverter;
+    }
+
+    public static PositionConverter createBy(String position) {
```
팩토리 클래스 메서드 이름으로 of도 많이 사용합니다~
PositionConverter.of()

Dae-Hwa replied:
그 동안 고민이 많았던 부분인데 피드백 주셔서 감사합니다. 앞으로도 자주 쓰일 것 같네요 😁
c189af1

LarryJung commented on src/main/java/kr/codesquad/freddie/chess/utils/PositionConverter.java:
```diff
@@ -0,0 +1,33 @@
+package kr.codesquad.freddie.chess.utils;
+
+public class PositionConverter {
```
string을 받아서 변환해주니 Converter가 맞긴한데.. Position이라고 이름을 지으면 좀 이상할까요?

Dae-Hwa replied:
말씀을 듣고 다시 생각해보니 독립적인 단위가 될 수 있겠다는 생각이 들었습니다. 변환 해주는 작업에만 너무 초점을 맞췄던 것 같아요. 수정하고 패키지도 board로 이동했습니다.
c189af1

LarryJung commented on src/test/java/kr/codesquad/freddie/chess/board/BoardTest.java:
```diff
@@ -20,6 +22,85 @@ void setBoard() {
+    void initializeWithSort() {
+        board.initializeWithSort();
+        Map<Color, List<Piece>> blackAndWhitePieces = board.getBlackAndWhitePieces();
+        assertThat(blackAndWhitePieces.get(Color.BLACK)).isEqualTo(new ArrayList<>(Arrays.asList(
```
new ArrayList<> 안해도 되지 않나요?? Arrays.asList만해도 List타입이 됩니다.

Dae-Hwa replied:
습관적으로 해버렸습니다. Arrays.asList만 해도 되겠네요 😅

Dae-Hwa replied:
68987ba

LarryJung commented on src/test/java/kr/codesquad/freddie/chess/utils/FileIndexConverterTest.java:
```diff
@@ -0,0 +1,30 @@
+
+class FileIndexConverterTest {
+    @Test
+    void create_outOfRange() {
```
네이밍 컨벤션을 지켜주세요~

Dae-Hwa replied:
`메소드명_테스트 조건`과 같은 방식으로 네이밍 했는데, 모두 카멜케이스로 작성하는게 맞는건가요?

LarryJung replied:
테스트코드 네이밍은 컨벤션을 관대하게 따르는 경향이 있는데요, 그래도 왠만하면 지켜주시고 필요하다면 @DisplayName 으로 대체하셔도 될 것 같네요~

Dae-Hwa replied:
알겠습니다. DisplayName을 좀 더 적극적으로 활용해봐야겠네요! 감사합니다
카멜 케이스로 수정은 9f5703f 에 커밋되어 있습니다.

LarryJung commented on src/main/java/kr/codesquad/freddie/chess/piece/Piece.java:
```diff
@@ -2,7 +2,7 @@
 import java.util.Objects;
 
-public class Piece {
+public class Piece implements Comparable<Piece> {
```
(참고) sort를 Stream으로 하셨으니 Comparable<Piece> 을 구현 안하고 하는 방법도 있습니다. 
stream sort 할때 Comparator를 사용할 수도 있습니다.

https://codechacha.com/ko/java8-stream-sorted/

Dae-Hwa replied:
감사합니다 참고해보겠습니다!

LarryJung reviewed (changes requested):
코드에 많은 고민이 느꼈습니다. ㅎㅎ 수고 많으셨습니다. 
개인적인 감상(?)으로는 그간 많이 공부하셔서 다양하게 적용해본다는 느낌이 들었구요~
객체 도출과정에서 나름의 의도가 드러난 것 같아서 좋았습니다. 다만 조금 장황하게 보여지는 부분도 있었지만 앞으로 좀 더 깔끔하고 확실하게 다듬어질 것이라 생각이 드네요~
제가 질문을 좀 많이 드렸는데, 보시고 고민! 부탁드립니다.^^

Dae-Hwa committed:
Refactor getNumberOf in File to use Wrapper instead type casting

(int) 캐스팅을 Long.valueOf(i).intValue()를 이용하여 수정

Dae-Hwa committed:
Fix getPieces and getCalculablePieces to use List instead Stream

Dae-Hwa committed:
Remove Calculable that  unnecessary abstraction for now

Dae-Hwa committed:
Rename TestBaseOfBoardPackage to BoardTestBase

Dae-Hwa committed:
Rename createBy in PositionConverter to of

Dae-Hwa committed:
Rename Converter classes to not use 'Converter' then move package

PositionConverter, FileIndexConverter, RankIndexConverter를 Position, FileIndex, RankIndex로 변경
변경 후 board패키지로 이동

Dae-Hwa committed:
Refactor BoardTest to use Arrays.asList only instead with ArrayList

Dae-Hwa commented:
리뷰 감사합니다! 나름대로 고민을 많이 했던 부분들을 알아봐주셔서 뿌듯?하네요 🥰
처리할 수 있는 부분들 처리하고, 피드백 주신 것들 중 이해가 안 되는 부분이나 궁금한 것들은 댓글로 질문을 남겼습니다.
감사합니다.

Dae-Hwa committed:
Rename test cases that don't follow camel case

Dae-Hwa committed:
Remove unnecessary import

Dae-Hwa committed:
Remove blackAndWihtePieces in Board

blackAndWihtePieces와 getter 메소드 제거. getPiecesBy로 정렬 된 기물들을 얻어낼 수 있다.

Dae-Hwa committed:
Remove unnecessary import

LarryJung reviewed (approved):
수고하셨습니다~ 다음 단계 진행하시죠!!
